spec/unit/pops/types/type_calculator_spec.rb

```diff
@@ -584,6 +590,42 @@ class Foo
       end
     end
 
+    context "for NotUndef, such that" do
+      it 'all types except types assignable from Undef are assignable to NotUndef' do
+        t = not_undef_t
+        tc = Puppet::Pops::Types::TypeCalculator.singleton
+        undef_t = Puppet::Pops::Types::PUndefType.new()
+        all_types().each do |c|
+          t2 = c.new
+          if tc.assignable?(t2, undef_t)
+            expect(t2).not_to be_assignable_to(t)
+          else
+            expect(t2).to be_assignable_to(t)
+          end
+        end
+      end
+
+      it 'all types assignable to T except types assignable from Undef are assignable to NotUndef[T]' do
+        t = not_undef_t(Puppet::Pops::Types::PDataType.new())
+        tc = Puppet::Pops::Types::TypeCalculator.singleton
+        undef_t = Puppet::Pops::Types::PUndefType.new()
+        all_types().select do |c|
+          t2 = c.new
+          if tc.assignable?(t.type, t2) && !tc.assignable?(t2, undef_t)
+            expect(t2).to be_assignable_to(t)
+          else
+            expect(t2).not_to be_assignable_to(t)
+          end
+        end
+      end
+
+      it 'NotUndef is not assignable to anything but NotUndef or Any' do
```

Comment:
it must be possible to assign T to NotUndef[T] (except if T is undef) and NotUndef[T] to T if the corresponding T's are assignable.